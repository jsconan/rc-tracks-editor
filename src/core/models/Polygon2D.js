/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { Vector2D } from './Vector2D.js';

/**
 * Represents a 2D polygon.
 */
export class Polygon2D {
    /**
     * Builds a polygon from the given cartesian coordinates.
     * @param {...Vector2D[]} points - The coordinates of the polygon.
     * @throws {TypeError} - If a point is not a Vector2D.
     */
    constructor(...points) {
        this.load(points.flat());
    }

    /**
     * The number of points in the polygon.
     * @type {number}
     */
    get length() {
        return this.points.length;
    }

    /**
     * Iterates over the points from the polygon.
     * @yields {*} - The next point of the polygon.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.points.values();
    }

    /**
     * Returns an iterator for the points from the polygon.
     * @yields {*} - The next point of the polygon.
     * @generator
     */
    *values() {
        yield* this.points.values();
    }

    /**
     * Applies a callback to each point of the polygon.
     * @param {polygonCallback} walker - A callback that will be applied to each point of the polygon.
     * @returns {List} - Chains the instance.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        this.points.forEach((point, index) => walker.call(this, point, index, this));

        return this;
    }

    /**
     * Maps the points of the polygon to an array.
     * @param {polygonCallback} mapper - A mapper callback that will be applied to each point of the polygon.
     * @returns {Array}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        return this.points.map((point, index) => mapper.call(this, point, index, this));
    }

    /**
     * Gets a point at a particular index.
     * @param {number} index - The index of the point to get
     * @returns {Vector2D} - The point at the given index. It may be undefined.
     */
    get(index) {
        return this.points[index];
    }

    /**
     * Sets a point at a particular index.
     * @param {number} index - The index where to set the point.
     * @param {Vector2D} point - The point to set at the index.
     * @returns {Polygon2D} - Chains the instance.
     * @throws {TypeError} - If the given point is not a Vector2D.
     * @throws {ReferenceError} - If the given index is out of bounds.
     */
    set(index, point) {
        if (index < 0 || index >= this.length) {
            throw new ReferenceError('The list index is out of bounds!');
        }

        Vector2D.validateInstance(point);

        this.points[index] = point;

        return this;
    }

    /**
     * Inserts points at a particular index.
     * @param {number} index - The index where to insert the points.
     * @param {...Vector2D|...Polygon2D} points - The points to insert at the index.
     * @returns {Polygon2D} - Chains the instance.
     * @throws {TypeError} - If the given point is not a Vector2D.
     */
    insert(index, ...points) {
        this.points.splice(index, 0, ...Polygon2D.extractPoints(points));

        return this;
    }

    /**
     * Inserts a point at a particular index from the given coordinates.
     * @param {number} index - The index where to insert the point.
     * @param {number} x - The X-coordinate of the point to insert.
     * @param {number} y - The Y-coordinate  of the point to insert.
     * @returns {Polygon2D} - Chains the instance.
     */
    insertPoint(index, x, y) {
        return this.insert(index, new Vector2D(x, y));
    }

    /**
     * Adds points at the end of the polygon.
     * @param {...Vector2D|...Polygon2D} points - The points to add.
     * @returns {Polygon2D} - Chains the instance.
     * @throws {TypeError} - If the given point is not a Vector2D.
     */
    add(...points) {
        this.points.push(...Polygon2D.extractPoints(points));

        return this;
    }

    /**
     * Inserts a point at the end of the polygon from the given coordinates.
     * @param {number} x - The X-coordinate of the point to insert.
     * @param {number} y - The Y-coordinate  of the point to insert.
     * @returns {Polygon2D} - Chains the instance.
     */
    addPoint(x, y) {
        return this.add(new Vector2D(x, y));
    }

    /**
     * Removes points from the given index.
     * @param {number} index - The index from where remove the points.
     * @param {number} count - The number of points to remove from the index.
     * @returns {number} - The number of deleted points.
     * @fires delete
     */
    delete(index, count = 1) {
        const removed = this.points.splice(index, count);

        return removed.length;
    }

    /**
     * Clears the polygon.
     * @returns {Polygon2D} - Chains the instance.
     */
    clear() {
        this.points = [];

        return this;
    }

    /**
     * Loads points from another source. The polygon is cleared before.
     * @param {*} iterator - An iterable object that can be used to set the points.
     * @returns {Polygon2D} - Chains the instance.
     * @throws {TypeError} - If a point is not a Vector2D.
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        this.points = Polygon2D.extractPoints(iterator);

        return this;
    }

    /**
     * Translates the polygon by the given vector.
     * @param {Vector2D} vector - The move to apply to the each point of the polygon.
     * @returns {Polygon2D} - Chains the instance.
     */
    translate(vector) {
        this.points = this.points.map(point => point.add(vector));

        return this;
    }

    /**
     * Rotates the polygon by the given angle around the origin (0, 0).
     * @param {number} angle - The rotation angle.
     * @returns {Polygon2D} - Chains the instance.
     */
    rotate(angle) {
        this.points = this.points.map(point => point.rotate(angle));

        return this;
    }

    /**
     * Rotates the polygon by the given angle around the given center.
     * @param {number} angle - The rotation angle.
     * @param {Vector2D} center - The rotation center.
     * @returns {Polygon2D} - Chains the instance.
     */
    rotateAround(angle, center) {
        this.points = this.points.map(point => point.rotateAround(angle, center));

        return this;
    }

    /**
     * Converts the polygon to a string.
     * @returns {string} - The string representation of the polygon.
     */
    toString() {
        return this.points.join(' ');
    }

    /**
     * Copy the polygon to an array.
     * @returns {Array} - An array containing the points from the polygon.
     */
    toArray() {
        return [...this.points];
    }

    /**
     * Extracts points from an iterator. It allows instances of Vector2D and Polygon2D.
     * Other values throw an error.
     * @param {*} points - An iterable object for listing the points.
     * @returns {Vector2D[]} - Returns a list of points.
     * @throws {TypeError} - If a point is not a Vector2D.
     */
    static extractPoints(points) {
        let list = [];
        for (let p of points) {
            if (p instanceof this) {
                list = [...list, ...p.points];
                continue;
            }

            if (Array.isArray(p)) {
                p = Vector2D.fromArray(p);
            }

            Vector2D.validateInstance(p);
            list.push(p);
        }

        return list;
    }

    /**
     * Validates that the given points are instances of Vector2D.
     * Otherwise, an error is thrown.
     * @param {*} points - An iterable object for listing the points.
     * @throws {TypeError} - If a point is not a Vector2D.
     */
    static validatePoints(points) {
        for (const p of points) {
            Vector2D.validateInstance(p);
        }
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * @typedef {import("./Vector2D.js").Vector2D} Vector2D
 */

/**
 * Callback called from the iteration algorithms.
 * @param {Vector2D} point - The current point being traversed.
 * @param {number} index - The index of the current point being traversed.
 * @returns {Vector2D} - Returns a point expected by the context.
 * @callback polygonCallback
 */
