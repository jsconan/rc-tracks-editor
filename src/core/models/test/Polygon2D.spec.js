/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Polygon2D } from '../Polygon2D.js';
import { Vector2D } from '../Vector2D.js';

const p1 = new Vector2D(10, 4);
const p2 = new Vector2D(10, 10);
const p3 = new Vector2D(4, 10);
const p4 = new Vector2D(0, 0);
const p5 = new Vector2D(-5, -6);
const p6 = new Vector2D(-10, 7);
const points = [p1, p2, p3, p4, p5, p6];

describe('Polygon2D', () => {
    it('is a class', () => {
        expect(Polygon2D).toEqual(expect.any(Function));
    });

    describe('can build a polygon', () => {
        it('from an empty list of points', () => {
            const polygon = new Polygon2D();

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual([]);
        });

        it('from a list of points given in an array', () => {
            const polygon = new Polygon2D(points);

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual(points);
        });

        it('from a list of points given as separated parameters', () => {
            const polygon = new Polygon2D(...points);

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual(points);
        });

        it('from a list of points given as a mix of Vector2D and Polygon2D', () => {
            const p = new Polygon2D(p3, p4, p5);
            const polygon = new Polygon2D(p1, p2, p, p6);

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual(points);
        });

        it('from a list of points given as separated arrays', () => {
            const polygon = new Polygon2D(points.slice(0, 2), points.slice(2));

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual(points);
        });

        it('from a list of points given as arrays and separated parameters', () => {
            const polygon = new Polygon2D(points.slice(0, 2), ...points.slice(2));

            expect(polygon).toBeInstanceOf(Polygon2D);
            expect(polygon.points).toStrictEqual(points);
        });
    });

    describe('throws an error', () => {
        it('if a point given at initialization is not a Vector2D', () => {
            expect(() => new Polygon2D(1)).toThrow('The object must be an instance of Vector2D!');
            expect(() => new Polygon2D({ x: 1, y: 2 })).toThrow('The object must be an instance of Vector2D!');
        });

        it('if a point to set is not a Vector2D', () => {
            const polygon = new Polygon2D(points);

            expect(() => polygon.set(0, 1, 2)).toThrow('The object must be an instance of Vector2D!');
            expect(() => polygon.set(0, { x: 1, y: 2 })).toThrow('The object must be an instance of Vector2D!');
        });

        it('if a point is attempted to be set at a wrong index', () => {
            const polygon = new Polygon2D(points);
            expect(() => polygon.set(-1, new Vector2D())).toThrow('The list index is out of bounds!');
            expect(() => polygon.set(6, new Vector2D())).toThrow('The list index is out of bounds!');
        });

        it('if a point to insert is not a Vector2D', () => {
            const polygon = new Polygon2D(points);

            expect(() => polygon.insert(0, 1, 2)).toThrow('The object must be an instance of Vector2D!');
            expect(() => polygon.insert(0, { x: 1, y: 2 })).toThrow('The object must be an instance of Vector2D!');
        });

        it('if a point to add is not a Vector2D', () => {
            const polygon = new Polygon2D(points);

            expect(() => polygon.add(1, 2)).toThrow('The object must be an instance of Vector2D!');
            expect(() => polygon.add({ x: 1, y: 2 })).toThrow('The object must be an instance of Vector2D!');
        });
    });

    it('has a length', () => {
        expect(new Polygon2D().length).toBe(0);
        expect(new Polygon2D(points).length).toBe(6);
    });

    it('implements the iteration protocol', () => {
        const polygon = new Polygon2D(points);

        expect(polygon[Symbol.iterator]).toEqual(expect.any(Function));
        expect(polygon[Symbol.iterator]()).not.toBe(polygon[Symbol.iterator]());
        expect([...polygon]).toStrictEqual(points);
    });

    it('can produce an iterator', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.values).toEqual(expect.any(Function));
        expect(polygon.values()[Symbol.iterator]).toEqual(expect.any(Function));

        expect(polygon.values()).not.toBe(polygon.values());
        expect([...polygon.values()]).toStrictEqual(points);
    });

    it('can walk over its points', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.forEach).toEqual(expect.any(Function));

        const iterator = points[Symbol.iterator]();
        let i = 0;
        const callback = jest.fn().mockImplementation(function (point, index, thisPolygon) {
            expect(this).toBe(polygon);
            expect(thisPolygon).toBe(polygon);
            expect(point).toBe(iterator.next().value);
            expect(index).toBe(i++);
        });

        expect(polygon.forEach(callback)).toBe(polygon);
        expect(callback).toHaveBeenCalledTimes(points.length);
    });

    it('needs a valid callback to walk over points', () => {
        expect(() => new Polygon2D().forEach()).toThrow('A callback function is expected!');
    });

    it('can map its points', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.map).toEqual(expect.any(Function));

        const iterator = points[Symbol.iterator]();
        let i = 0;
        const callback = jest.fn().mockImplementation(function (point, index, thisPolygon) {
            expect(this).toBe(polygon);
            expect(thisPolygon).toBe(polygon);
            expect(point).toBe(iterator.next().value);
            expect(index).toBe(i++);
            return point.mulScalar(10);
        });

        const mapped = polygon.map(callback);
        expect(mapped).not.toBe(polygon);
        expect(mapped).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(points.length);
    });

    it('needs a valid callback to map points', () => {
        expect(() => new Polygon2D().map()).toThrow('A callback function is expected!');
    });

    it('can get a point from a particular index', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.get).toEqual(expect.any(Function));

        expect(polygon.get(0)).toBe(points[0]);
        expect(polygon.get(1)).toBe(points[1]);
        expect(polygon.get(2)).toBe(points[2]);
        expect(polygon.get(3)).toBe(points[3]);
        expect(polygon.get(4)).toBe(points[4]);
        expect(polygon.get(5)).toBe(points[5]);
        expect(polygon.get(6)).toBeUndefined();
    });

    it('can set a point to a particular index', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.set).toEqual(expect.any(Function));

        expect(polygon.set(0, new Vector2D(4, 5))).toBe(polygon);
        expect(polygon.set(1, new Vector2D(18, 15))).toBe(polygon);
        expect(polygon.set(2, new Vector2D(42, 5))).toBe(polygon);
        expect([...polygon]).toMatchSnapshot();
    });

    describe('can insert points', () => {
        describe('from existing vectors', () => {
            it('at a particular index', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(1, new Vector2D(1, 2), new Vector2D(3, 4))).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the beginning', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(0, new Vector2D(1, 2), new Vector2D(3, 4))).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the end', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(6, new Vector2D(1, 2), new Vector2D(3, 4))).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });
        });

        describe('from polygons', () => {
            it('at a particular index', () => {
                const p = new Polygon2D(p4, p5);
                const polygon = new Polygon2D(p1, p2, p3);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(1, p, p6)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the beginning', () => {
                const p = new Polygon2D(p4, p5);
                const polygon = new Polygon2D(p1, p2, p3);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(0, p, p6)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the end', () => {
                const p = new Polygon2D(p4, p5);
                const polygon = new Polygon2D(p1, p2, p3);

                expect(polygon.insert).toEqual(expect.any(Function));

                expect(polygon.insert(3, p, p6)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });
        });

        describe('from coordinates', () => {
            it('at a particular index', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insertPoint).toEqual(expect.any(Function));

                expect(polygon.insertPoint(1, 1, 2)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the beginning', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insertPoint).toEqual(expect.any(Function));

                expect(polygon.insertPoint(0, 1, 2)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });

            it('at the end', () => {
                const polygon = new Polygon2D(points);

                expect(polygon.insertPoint).toEqual(expect.any(Function));

                expect(polygon.insertPoint(4, 1, 2)).toBe(polygon);
                expect([...polygon]).toMatchSnapshot();
            });
        });
    });

    describe('can add points to the polygon', () => {
        it('from existing vectors', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.add).toEqual(expect.any(Function));

            expect(polygon.add(new Vector2D(1, 2), new Vector2D(3, 4))).toBe(polygon);
            expect([...polygon]).toMatchSnapshot();
        });

        it('from polygon', () => {
            const p = new Polygon2D(p4, p5);
            const polygon = new Polygon2D(p1, p2, p3);

            expect(polygon.add).toEqual(expect.any(Function));

            expect(polygon.add(p, p6)).toBe(polygon);
            expect([...polygon]).toMatchSnapshot();
        });

        it('from coordinates', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.addPoint).toEqual(expect.any(Function));

            expect(polygon.addPoint(1, 2)).toBe(polygon);
            expect([...polygon]).toMatchSnapshot();
        });
    });

    it('can remove points from a particular index', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.delete).toEqual(expect.any(Function));

        expect(polygon.delete(4, 2)).toBe(2);
        expect(polygon.delete(4)).toBe(0);
        expect([...polygon]).toMatchSnapshot();
    });

    it('can clear the polygon', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.clear).toEqual(expect.any(Function));

        expect([...polygon]).toEqual(points);
        expect(polygon.clear()).toBe(polygon);
        expect([...polygon]).toEqual([]);
    });

    it('can load points from another source', () => {
        const polygon = new Polygon2D();

        expect(polygon.load).toEqual(expect.any(Function));

        expect(polygon.load({})).toBe(polygon);
        expect([...polygon]).toEqual([]);

        expect(
            polygon.load([
                [1, 2],
                [3, 4]
            ])
        ).toBe(polygon);
        expect([...polygon]).toMatchSnapshot();

        expect(polygon.load(points)).toBe(polygon);
        expect([...polygon]).toEqual(points);

        const p = new Polygon2D();
        expect(p.load(polygon)).toBe(p);
        expect([...p]).toEqual(points);
    });

    it('can translate the points', () => {
        const polygon = new Polygon2D(points);

        expect(polygon.translate).toEqual(expect.any(Function));

        expect(polygon.translate(new Vector2D(5, 8))).toMatchSnapshot();
    });

    describe('can rotate the points', () => {
        it('around the origin', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.rotate).toEqual(expect.any(Function));

            expect(polygon.rotate(30)).toMatchSnapshot();
        });

        it('around the given center', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.rotateAround).toEqual(expect.any(Function));

            expect(polygon.rotateAround(30, new Vector2D(5, 8))).toMatchSnapshot();
        });
    });

    describe('can export the points', () => {
        it('to a string', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.toString).toEqual(expect.any(Function));
            expect(polygon.toString()).toMatchSnapshot();
        });

        it('to an array', () => {
            const polygon = new Polygon2D(points);

            expect(polygon.toArray).toEqual(expect.any(Function));
            expect(polygon.toArray()).toEqual(points);
        });
    });

    it('extracts points from an iterator', () => {
        const polygon = new Polygon2D(p3, p4, p5);

        expect(Polygon2D.extractPoints([])).toStrictEqual([]);
        expect(Polygon2D.extractPoints([p1, p2, p3])).toStrictEqual([p1, p2, p3]);
        expect(Polygon2D.extractPoints([p1, p2, polygon, p6])).toStrictEqual([p1, p2, p3, p4, p5, p6]);

        expect(() => Polygon2D.extractPoints({})).toThrow('points is not iterable');
        expect(() => Polygon2D.extractPoints([{}])).toThrow('The object must be an instance of Vector2D!');
    });

    it('can validate a list contains valid points', () => {
        expect(() => Polygon2D.validatePoints(points)).not.toThrow();
        expect(() => Polygon2D.validatePoints([{}])).toThrow('The object must be an instance of Vector2D!');
    });

    it('can validate an object is an instance of the class', () => {
        const polygon = new Polygon2D();
        expect(() => Polygon2D.validateInstance(polygon)).not.toThrow();
        expect(() => Polygon2D.validateInstance({})).toThrow('The object must be an instance of Polygon2D!');
    });
});
