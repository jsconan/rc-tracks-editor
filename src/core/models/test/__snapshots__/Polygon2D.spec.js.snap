// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polygon2D can add points to the polygon from coordinates 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
]
`;

exports[`Polygon2D can add points to the polygon from existing vectors 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 3,
    "y": 4,
  },
]
`;

exports[`Polygon2D can add points to the polygon from polygon 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can export the points to a string 1`] = `"10,4 10,10 4,10 0,0 -5,-6 -10,7"`;

exports[`Polygon2D can insert points from coordinates at a particular index 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can insert points from coordinates at the beginning 1`] = `
[
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can insert points from coordinates at the end 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can insert points from existing vectors at a particular index 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 3,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can insert points from existing vectors at the beginning 1`] = `
[
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 3,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can insert points from existing vectors at the end 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 3,
    "y": 4,
  },
]
`;

exports[`Polygon2D can insert points from polygons at a particular index 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
]
`;

exports[`Polygon2D can insert points from polygons at the beginning 1`] = `
[
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
]
`;

exports[`Polygon2D can insert points from polygons at the end 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can load points from another source 1`] = `
[
  Vector2D {
    "x": 1,
    "y": 2,
  },
  Vector2D {
    "x": 3,
    "y": 4,
  },
]
`;

exports[`Polygon2D can map its points 1`] = `
[
  Vector2D {
    "x": 100,
    "y": 40,
  },
  Vector2D {
    "x": 100,
    "y": 100,
  },
  Vector2D {
    "x": 40,
    "y": 100,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -50,
    "y": -60,
  },
  Vector2D {
    "x": -100,
    "y": 70,
  },
]
`;

exports[`Polygon2D can remove points from a particular index 1`] = `
[
  Vector2D {
    "x": 10,
    "y": 4,
  },
  Vector2D {
    "x": 10,
    "y": 10,
  },
  Vector2D {
    "x": 4,
    "y": 10,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Polygon2D can rotate the points around the given center 1`] = `
Polygon2D {
  "points": [
    Vector2D {
      "x": 11.330127018922195,
      "y": 7.035898384862245,
    },
    Vector2D {
      "x": 8.330127018922195,
      "y": 12.232050807568877,
    },
    Vector2D {
      "x": 3.1339745962155616,
      "y": 9.232050807568877,
    },
    Vector2D {
      "x": 4.6698729810778055,
      "y": -1.4282032302755088,
    },
    Vector2D {
      "x": 3.339745962155612,
      "y": -9.124355652982143,
    },
    Vector2D {
      "x": -7.49038105676658,
      "y": -0.3660254037844375,
    },
  ],
}
`;

exports[`Polygon2D can rotate the points around the origin 1`] = `
Polygon2D {
  "points": [
    Vector2D {
      "x": 6.660254037844387,
      "y": 8.464101615137753,
    },
    Vector2D {
      "x": 3.660254037844388,
      "y": 13.660254037844386,
    },
    Vector2D {
      "x": -1.5358983848622443,
      "y": 10.660254037844387,
    },
    Vector2D {
      "x": 0,
      "y": 0,
    },
    Vector2D {
      "x": -1.330127018922194,
      "y": -7.696152422706632,
    },
    Vector2D {
      "x": -12.160254037844387,
      "y": 1.0621778264910722,
    },
  ],
}
`;

exports[`Polygon2D can set a point to a particular index 1`] = `
[
  Vector2D {
    "x": 4,
    "y": 5,
  },
  Vector2D {
    "x": 18,
    "y": 15,
  },
  Vector2D {
    "x": 42,
    "y": 5,
  },
  Vector2D {
    "x": 0,
    "y": 0,
  },
  Vector2D {
    "x": -5,
    "y": -6,
  },
  Vector2D {
    "x": -10,
    "y": 7,
  },
]
`;

exports[`Polygon2D can translate the points 1`] = `
Polygon2D {
  "points": [
    Vector2D {
      "x": 15,
      "y": 12,
    },
    Vector2D {
      "x": 15,
      "y": 18,
    },
    Vector2D {
      "x": 9,
      "y": 18,
    },
    Vector2D {
      "x": 5,
      "y": 8,
    },
    Vector2D {
      "x": 0,
      "y": 2,
    },
    Vector2D {
      "x": -5,
      "y": 15,
    },
  ],
}
`;
